### PR TITLE
Add label for generic arg (+ APIT) and RPIT callables in `label_fn_like`

### LINKED ISSUE #98308: Don't point to `Fn::call` when the arguments passed to a closure are incorrect
Given the following code: 

```rust
fn takes_callback(callback: impl Fn(i32)) {
    callback(&0);
}
```

The current output is dependent on whether you have the `rust-src` component installed. With it installed:

```
error[E0308]: mismatched types
  --> src/main.rs:2:14
   |
2  |     callback(&0);
   |     -------- ^^ expected `i32`, found `&{integer}`
   |     |
   |     arguments to this function are incorrect
   |
note: associated function defined here
  --> /home/jnelson/.local/lib/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:77:27
   |
77 |     extern "rust-call" fn call(&self, args: Args)...
   |                           ^^^^
help: consider removing the borrow
   |
2  -     callback(&0);
2  +     callback(0);
```

without it (e.g. on playground), the same except the note looks very strange because it doesn't actually point anywhere:
```
note: associated function defined here
help: consider removing the borrow
```

Ideally the output should look like:

```
error[E0308]: mismatched types
  --> src/main.rs:2:14
   |
2  |     callback(&0);
   |     -------- ^^ expected `i32`, found `&{integer}`
   |     |
   |     arguments to this function are incorrect
   |
note: function defined here:
  --> src/main.rs:1:x
   |
1 |     fn takes_callback(callback: impl Fn(i32)) {
   |                                ^^^^^^^^^^^^
help: consider removing the borrow
   |
2  -     callback(&0);
2  +     callback(0);
```

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #98308

### CODE REVIEW

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs [hunk 1]
```diff
@@ -1783,47 +1802,86 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                                err.span_note(span, "callable defined here");
+                                return;
+                            }
+                        }
+                    }
+                }
+                ty::Opaque(new_def_id, _) | ty::Closure(new_def_id, _) | ty::FnDef(new_def_id, _) => {
+                    def_id = new_def_id;
+                }
+                _ => {
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:1856]: What about types that have a manual impl of `FnOnce`?  Should we add something for them?

**Review by cjgillot — COMMENTED** [review]
Thanks @compiler-errors.  This looks good.  I left a few questions.

File: src/test/ui/mismatched_types/overloaded-calls-bad.stderr [hunk 2]
```diff
@@ -5,24 +5,13 @@ LL |     let ans = s("what");
    |               - ^^^^^^ expected `isize`, found `&str`
    |               |
    |               arguments to this function are incorrect
-   |
-note: associated function defined here
-  --> $SRC_DIR/core/src/ops/function.rs:LL:COL
-   |
-LL |     extern "rust-call" fn call_mut(&mut self, args: Args) -> Self::Output;
-   |                           ^^^^^^^^
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply src/test/ui/mismatched_types/overloaded-calls-bad.stderr:13]: It's a bit of work to do so, but done!

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs [hunk 3]
```diff
@@ -1783,47 +1799,125 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        }
+                    }
+                }
+                ty::Opaque(new_def_id, _)
+                | ty::Closure(new_def_id, _)
+                | ty::FnDef(new_def_id, _) => {
+                    def_id = new_def_id;
+                }
+                _ => {
+                    let new_def_id = self.probe(|_| {
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:1861]:
```suggestion
                    // Look for a user-provided impl of a `Fn` trait, and point to it.
                    let new_def_id = self.probe(|_| {
```

**Review by cjgillot — APPROVED** [review]
r=me with an extra comment.

**Comment by jyn514** [comment]
This is amazing, thank you so much!!

**Pull request merged**